packages/devextreme/testing/tests/DevExpress.ui.widgets/tabs.tests.js

```diff
@@ -422,6 +423,200 @@ QUnit.module('General', () => {
+            this.CANCEL_DELAY = 0;
+        },
+        afterEach() {
+            this.clock.restore();
+        }
+    }, () => {
+        QUnit.test('Selection should remain unchanged when onSelectionChanging is cancelled initially', function(assert) {
+            let count = 0;
+            const onSelectionChangingSpy = sinon.spy(function(e) {
+                e.cancel = true;
+                count++;
+            });
+            const onSelectionChangedSpy = sinon.spy();
+            this.reinit({
+                onSelectionChanging: onSelectionChangingSpy,
+                onSelectionChanged: onSelectionChangedSpy,
+            });
+
+            const $item = this.$tabs.find(`.${TABS_ITEM_CLASS}`).eq(1);
+            $item.trigger('dxclick');
+
+            assert.strictEqual(count, 2, 'onSelectionChanging should be called twice');
+            assert.strictEqual(onSelectionChangedSpy.callCount, 0, 'onSelectionChanged should not be called');
+            assert.strictEqual(this.tabInstance.option('selectedIndex'), 0, 'selectedIndex should remain 0 when cancel is true');
+            assert.deepEqual(this.tabInstance.option('selectedItem'), { text: '0' }, 'selectedItem should remain the first item');
+            assert.deepEqual(this.tabInstance.option('selectedItems'), [{ text: '0' }], 'selectedItems should remain the first item');
+            assert.deepEqual(this.tabInstance.option('selectedItemKeys'), [{ text: '0' }], 'selectedItemKeys should remain the first key');
+        });
+
+        QUnit.test('Selection should be cancelled synchronously when e.cancel is set to true', function(assert) {
+            const onSelectionChangingSpy = sinon.spy(function(e) {
+                e.cancel = true;
+            });
+
+            this.reinit({
+                onSelectionChanging: onSelectionChangingSpy,
+            });
+            this.$item = this.$tabs.find(`.${TABS_ITEM_CLASS}`).eq(1);
+            this.$item.trigger('dxclick');
+
+            assert.strictEqual(onSelectionChangingSpy.callCount, 2, 'onSelectionChanging should be called');
+            assert.strictEqual(this.onSelectionChangedSpy.callCount, 0, 'onSelectionChanged should not be called');
+            assert.ok(onSelectionChangingSpy.getCall(0).args[0].cancel, 'e.cancel should be set to true');
+            assert.strictEqual(this.tabInstance.option('selectedIndex'), 0, 'selectedIndex should remain 0 when cancelled');
+            assert.deepEqual(this.tabInstance.option('selectedItem'), { text: '0' }, 'selectedItem should remain the first item');
+            assert.deepEqual(this.tabInstance.option('selectedItems'), [{ text: '0' }], 'selectedItems should remain the first item');
+            assert.deepEqual(this.tabInstance.option('selectedItemKeys'), [{ text: '0' }], 'selectedItemKeys should remain the first key');
+        });
+
+        QUnit.test('Selection should not change when onSelectionChanging is cancelled with a rejected promise', function(assert) {
```

Comment:
I believe that promise rejection is the same case as promise resolving with false